memory_allocator.c

```diff
@@ -16,9 +16,16 @@
 TAILQ_HEAD(mb_list, mem_block);
 TAILQ_HEAD(mp_arena, mem_arena);
 
+int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
 typedef struct mem_block {
   uint32_t mb_magic; /* if overwritten report a memory corruption error */
-  uint32_t mb_size;
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
```

Comment:
You are right. With the opposite, I have to use `abs` only once, and the calculations are simpler too.